x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/cache/blob/BlobStoreCacheMaintenanceService.java

```diff
@@ -236,4 +353,285 @@ public void onFailure(Exception e) {
+
+                if (searchResponse == null) {
+                    final SearchSourceBuilder searchSource = new SearchSourceBuilder();
+                    searchSource.trackScores(false);
+                    searchSource.sort("_shard_doc");
+                    searchSource.size(batchSize);
+                    final PointInTimeBuilder pointInTime = new PointInTimeBuilder(pitId);
+                    searchSource.pointInTimeBuilder(pointInTime);
+                    pointInTime.setKeepAlive(keepAlive);
+                    final SearchRequest searchRequest = new SearchRequest();
+                    searchRequest.source(searchSource);
+                    if (searchAfter != null) {
+                        searchSource.searchAfter(searchAfter);
+                    }
+                    clientWithOrigin.execute(SearchAction.INSTANCE, searchRequest, new ActionListener<>() {
+                        @Override
+                        public void onResponse(SearchResponse response) {
+                            maintenanceTask.total.compareAndSet(0L, response.getHits().getTotalHits().value);
+                            maintenanceTask.searchResponse = response;
+                            maintenanceTask.searchAfter = null;
+                            executeNext(maintenanceTask);
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            complete(maintenanceTask, e);
+                        }
+                    });
+                    return;
+                }
+
+                final SearchHit[] searchHits = searchResponse.getHits().getHits();
+                if (searchHits != null && searchHits.length > 0) {
+                    final ClusterState state = clusterService.state();
+                    final BulkRequest bulkRequest = new BulkRequest();
+
+                    final Set<Tuple<String, String>> knownSnapshots = new HashSet<>();
+                    final Set<Tuple<String, String>> missingSnapshots = new HashSet<>();
+                    final Set<String> knownRepositories = state.metadata()
+                        .custom(RepositoriesMetadata.TYPE, RepositoriesMetadata.EMPTY)
+                        .repositories()
+                        .stream()
+                        .map(RepositoryMetadata::name)
+                        .collect(Collectors.toSet());
+
+                    Object[] lastSortValues = null;
+                    for (SearchHit searchHit : searchHits) {
+                        assert searchHit.getId() != null;
+                        assert searchHit.hasSource();
+
+                        // See {@link BlobStoreCacheService#generateId}
+                        // doc id = {repository name}/{snapshot id}/{snapshot index id}/{shard id}/{file name}/@{file offset}
+                        final String[] parts = Objects.requireNonNull(searchHit.getId()).split("/");
```

Comment:
I noticed that the repository name is part of the document id :( Using our own maintenance task here could help to reindex the docs without the repository name, if we think the snapshot UUID + index UUID are unique enough to avoid collisions.